src/tdigest.rs

```diff
@@ -0,0 +1,269 @@
+        let lerp_val = (v - self.centroids[index - 1].mean) / (self.centroids[index].mean - self.centroids[index - 1].mean);
+        quantile_l + lerp_val * (quantile_r - quantile_l)
+    }
+
+    // shamefully stolen from the better tdigest rust crate
+    /// Estimates the value at the given quantile.
+    pub fn estimate_value(&self, q: f32) -> f32 {
+        if self.centroids.len() == 0 {
+            return 0.0;
+        }
+
+        let count_ = self.centroids.iter().fold(0, |sum, x| sum + x.weight) as f32;
+        let rank = count_ * q;
+
+        let self_min = self.centroids[0].mean;
+        let self_max = self.centroids.last().expect("uh oh").mean;
+
+        // the following chunk of code does the same thing as search_centroids but it's based on quantile instead of mean and also it's much faster for the use case (most queries are going to be at the extremes, so linear search works out to be pretty fast)
+        let mut pos: usize;
+        let mut t: f32;
+        if q > 0.5 {
+            if q >= 1.0 {
+                return self_max;
+            }
+
+            pos = 0;
+            t = count_;
+
+            for (k, centroid) in self.centroids.iter().enumerate().rev() {
+                t -= centroid.weight as f32;
+
+                if rank >= t {
+                    pos = k;
+                    break;
+                }
+            }
+        } else {
+            if q <= 0.0 {
+                return self_min;
+            }
+
+            pos = self.centroids.len() - 1;
+            t = 0.0;
+
+            for (k, centroid) in self.centroids.iter().enumerate() {
+                if rank < t + centroid.weight as f32 {
+                    pos = k;
+                    break;
+                }
+
+                t += centroid.weight as f32;
+            }
+        }
+
+        // TODO: document this lmao
+        let mut delta = 0.0;
+        let mut min: f32 = self_min.clone();
+        let mut max: f32 = self_max.clone();
+
+        if self.centroids.len() > 1 {
+            if pos == 0 {
+                delta = self.centroids[pos + 1].mean - self.centroids[pos].mean;
+                max = self.centroids[pos + 1].mean;
+            } else if pos == (self.centroids.len() - 1) {
+                delta = self.centroids[pos].mean - self.centroids[pos - 1].mean;
+                min = self.centroids[pos - 1].mean;
+            } else {
+                delta = (self.centroids[pos + 1].mean - self.centroids[pos - 1].mean) / 2.0;
+                min = self.centroids[pos - 1].mean;
+                max = self.centroids[pos + 1].mean;
+            }
+        }
+
+        let value = self.centroids[pos].mean + ((rank - t) / self.centroids[pos].weight as f32 - 0.5) * delta;
+        Self::clamp(value, min, max)
```

Comment:
Since you are only using this once here, as suggested above, you could write `value.max(min).min(max)` :smiling_face_with_tear: